[CL-2535] Crop project/folder card images

alexander-cit opened the pull request:
## ToDo refactoring

### folders
- [ ] get rid of `mode === 'new'` in `onSubmit`

### projects
- [x] combine `save` and `saveForm`

### both
- [x] use one object instead of array `projectImages` (do not touch BE)
- [x] use one object instead of array `projectFolderImages` (do not touch BE)
- [x] use the same names for the same things on project and folder forms
  - [x] `onSubmit` - `saveForm`
  - [x] `setStatus` - `setSubmitState`
- [ ] tie both forms so that in the future changes in one will require changes in another

alexander-cit committed:
[CL-2535] Make component name the same as folder name

alexander-cit committed:
[CL-2535] Crop project/folder card images

Commit:
Translations updated by CI (extract-intl)

alexander-cit commented:
@brentguf if you have time, you're more than welcome to take a look at the draft and especially at my refactoring plan.

P.S. The code works (didn't test edge cases though).

brentguf commented on front/app/containers/Admin/projectFolders/containers/settings/ProjectFolderForm.tsx:
```diff
@@ -180,6 +185,10 @@ const ProjectFolderForm = ({ mode, projectFolderId }: Props) => {
+  const handleProjectFolderImagesOnRemoveNoArgs = useCallback(() => {
+    handleProjectFolderImageOnRemove(projectFolderImages[0]);
+  }, [projectFolderImages]);
+
```
I don't think this useCallback does something useful (nor is needed). Also, in this file many functions are wrapped with useCallback and an empty dependencies array, which is the same as not wrapping with useCallback at all. So let's start cleaning up this file by removing all these?

brentguf replied:
In fact, no function in this file needs useCallback. They can all be removed.

brentguf commented on front/app/containers/Admin/projectFolders/containers/settings/ProjectFolderForm.tsx:
```diff
@@ -261,9 +270,14 @@ const ProjectFolderForm = ({ mode, projectFolderId }: Props) => {
+                      projectFolderImagesBase64
+                    )
+                  )
+                : [];
```
You're not using file anymore, so why would you map over projectFolderImages?

brentguf commented on front/app/containers/Admin/projectFolders/containers/settings/ProjectFolderForm.tsx:
```diff
@@ -288,11 +302,16 @@ const ProjectFolderForm = ({ mode, projectFolderId }: Props) => {
+                      projectFolderImagesBase64
+                    )
+                  )
+              : [];
```
Same question here.

brentguf commented on front/app/containers/Admin/projects/project/general/index.tsx:
```diff
@@ -1,4 +1,4 @@
+import React, { useState, useEffect, useCallback } from 'react';
 import { Multiloc, UploadFile } from 'typings';
 import { isEmpty, get, isString } from 'lodash-es';
 import CSSTransition from 'react-transition-group/CSSTransition';
```
Same remarks in this file as in the folder form.

brentguf commented:
```
folders
 get rid of mode === 'new' in onSubmit
```

This is a pattern we have in the FE that you ideally check with Iva on how to change.

```
projects
 combine save and saveForm
```
👍 

These forms are also missing tests, so it could be useful to add some before refactoring, but we don't want this epic to blow up.

```
both
 use one object instead of arrays projectImages / projectFolderImages (do not touch BE)
 use the same names for the same things on project and folder forms
 onSubmit - saveForm
 setStatus - setSubmitState
```
Agreed on these.

` tie both forms so that in the future changes in one will require changes in another`
Careful here. Again, I would check with Iva on this.

alexander-cit committed:
[CL-2535] Remove redundant save function

These two examples behave the same way:

```js
const f = async () => { return await fetch('http://localhost:3000/web_api/v1/app_configuration') }
f().then(data => console.log(data))

const f2 = () => { return fetch('http://localhost:3000/web_api/v1/app_configuration') }
f2().then(data => console.log(data))
```

alexander-cit committed:
[CL-2535] Use same names for same things on project and folder forms

Project form used for this refactoring is
front/app/components/admin/SubmitWrapper/index.tsx

alexander-cit committed:
[CL-2535] Keep one image instead of array for project card

alexander-cit committed:
[CL-2535] Remove all unused project and folder images

alexander-cit committed:
[CL-2535] Fix lint errors

alexander-cit commented:
[This images number issue](https://citizenlabco.slack.com/archives/C015M14HYSF/p1674228018666059) means that `projectImages` and `projectImagesToRemove` could be refactored only together. If only one image is stored in `projectImages`, the rest will not be removed when the component is mounted. It means that a new image will be added to the end of the list, and the old one will be removed from the beginning of the list. So, the new image won't be visible.

It also means that we should fix the images number issue on the BE. Hence the migration is added.

alexander-cit committed:
[CL-2535] Keep one image instead of array for folder card

alexander-cit committed:
Merge branch 'CL-2534-project-folder-card-images' into CL-2535-crop-project-card

alexander-cit committed:
Merge branch 'CL-2534-project-folder-card-images' into CL-2535-crop-project-card

alexander-cit committed:
[CL-2535] Remove console.log

alexander-cit committed:
[CL-2535] Use constant CARD_IMAGE_ASPECT_RATIO consistently

alexander-cit commented on front/app/containers/Admin/projects/project/general/index.tsx:
```diff
@@ -452,6 +453,8 @@ const AdminProjectsProjectGeneral = () => {
     !isNilOrError(project) ? project : null
   );
 
+  const projectCardShouldBeSaved = projectCard ? !projectCard.remote : false;
```
But I use just the word `card` (w/o `image`) in all other variables. Shouldn't it be consistent?

brentguf commented on front/app/containers/Admin/projects/project/general/index.tsx:
```diff
@@ -31,6 +31,7 @@ import {
   ParticipationContextWrapper,
```
Could we change projectCard to projectCardImage in the entire file?

brentguf replied:
The card currently means the whole card, also the part under the image. So this renaming is important to avoid confusion.

brentguf commented on front/app/containers/Admin/projects/project/general/index.tsx:
```diff
@@ -290,24 +302,22 @@ const AdminProjectsProjectGeneral = () => {
-              return addProjectImage(latestProjectId, file.base64);
-            }
+        const cardToAddPromise =
+          croppedProjectCardBase64 &&
```
Could you add this in a variable with a good name to indicate what you're checking here?

brentguf replied:
Isn't just having the croppedProjectCardBase64 enough of a check that there's a new image? The fact that you're only using this value in the function call, not the other two (folderCardImage and !folderCardImage.remote), also indicates that the other two are probably not needed.

brentguf replied:
So we could do 

```
const cardImageToAddPromise =
          croppedProjectCardBase64 && latestProjectId
            ? addProjectImage(latestProjectId, croppedProjectCardBase64)
            : null;
```

instead of 


```
        const cardImageWasChanged =
          croppedProjectCardBase64 &&
          projectCardImage &&
          !projectCardImage.remote;
        const cardImageToAddPromise =
        croppedProjectCardBase64 && latestProjectId
            ? addProjectImage(latestProjectId, croppedProjectCardBase64)
            : null;
```

brentguf commented on front/app/containers/Admin/projects/project/general/index.tsx:
```diff
@@ -565,12 +569,22 @@ const AdminProjectsProjectGeneral = () => {
+          </Box>
+        ) : (
+          <ProjectImageDropzone
+            projectImages={projectCard && [projectCard]}
```
`projectCard ? [projectCard] : null` states the intention better imo. Wdyt?

alexander-cit replied:
It's the same for me.

brentguf replied:
With my version you see what the value will be set to without having to hover projectCard to see potential values. So not the same imo.

brentguf commented on front/app/containers/Admin/projectFolders/containers/settings/ProjectFolderForm.tsx:
```diff
@@ -261,24 +265,28 @@ const ProjectFolderForm = ({ mode, projectFolderId }: Props) => {
-                addProjectFolderImage(projectFolder.id, file.base64)
-              );
+              const cardToAddPromise =
+                croppedFolderCardBase64 && folderCard && !folderCard.remote
```
The same remark here: could you make this a variable to clarify the intention?

brentguf commented on front/app/containers/Admin/projectFolders/containers/settings/ProjectFolderForm.tsx:
```diff
@@ -389,8 +400,12 @@ const ProjectFolderForm = ({ mode, projectFolderId }: Props) => {
     return null;
   }
 
+  const projectFolderImagesShouldBeSaved = folderCard
```
projectFolderImageShouldBeSaved instead (singular)?

brentguf reviewed (changes requested):
Some suggestions to increase readability.

A blocking issue though: when I upload an image, the title 'Project card image' disappears:

<img width="331" alt="Screenshot 2023-01-23 at 17 33 34" src="https://user-images.githubusercontent.com/16427929/214095918-5c3a85cf-5932-4ead-a039-e326762fa6af.png">

<img width="586" alt="Screenshot 2023-01-23 at 17 33 08" src="https://user-images.githubusercontent.com/16427929/214096679-a3511e07-f23d-4d98-9520-51b6fc785071.png">

Also, the preview is problematic. It should be bigger; it's tiny atm:

<img width="350" alt="Screenshot 2023-01-23 at 17 36 23" src="https://user-images.githubusercontent.com/16427929/214096233-13aa92be-27bd-4d51-bf25-fd767dc424a8.png">

Lastly, I think it's a problem that the dimensions of the preview don't show what it will look like when they go search the project card (on a desktop). We should indicate the dimensions/cropper are for mobile, and then convey somehow what it will look like on desktop. (Reuse the preview component for homepage banner?)

alexander-cit committed:
[CL-2535] Always show card image header

alexander-cit commented:
> Lastly, I think it's a problem that the dimensions of the preview don't show what it will look like when they go search the project card (on a desktop). We should indicate the dimensions/cropper are for mobile, and then convey somehow what it will look like on desktop.

I believe it will be solved by https://citizenlab.atlassian.net/browse/CL-2672

brentguf commented:
> > Lastly, I think it's a problem that the dimensions of the preview don't show what it will look like when they go search the project card (on a desktop). We should indicate the dimensions/cropper are for mobile, and then convey somehow what it will look like on desktop.
> 
> I believe it will be solved by https://citizenlab.atlassian.net/browse/CL-2672

I think that's too hard to understand. A preview > a 1000 words :)

brentguf committed:
Extract SelectPreview

Commit:
Translations updated by CI (extract-intl)

brentguf committed:
Use 'messages'

brentguf committed:
Merge branch 'crop-project-card-preview' of https://github.com/CitizenLabDotCo/citizenlab into crop-project-card-preview

Commit:
Translations updated by CI (extract-intl)

brentguf committed:
Move TPreviewDevice to SelectPreviewDevice

brentguf committed:
Remove unneeded styles

brentguf committed:
Show preview for project card image

brentguf committed:
Order of devices

brentguf committed:
Merge branch 'crop-project-card-preview' of https://github.com/CitizenLabDotCo/citizenlab into crop-project-card-preview

brentguf committed:
ProjectFolderCardImageDropzone

brentguf committed:
Simplify naming

brentguf committed:
Preview device is phone on image removal

brentguf committed:
Revert "Translations updated by CI (extract-intl)"

This reverts commit ed8dc40f3fb3ca34be38482bfc410ce3c59da55c.

brentguf committed:
Move SelectPreviewDevice to admin/components

brentguf committed:
Use absolute path

brentguf committed:
Merge pull request #3746 from CitizenLabDotCo/crop-project-card-preview

CL-2677 Preview functionality for project/folder card images

alexander-cit commented:
@jinjagit The task I was talking about is here. It would be nice if you could have a look.

jinjagit reviewed (approved):
back/lib/tasks/data_migrate/20220123_remove_unused_project_images.rake looks good to me. 👍

I would want to run it locally, probably on a copy of a tenant that already has multiple images for either a Project & or Folder, and check the Rails logs to see what transactions get committed, etc. You may well have already done that, though.

alexander-cit commented:
@jinjagit Sure, I did with some manually created extra images.

alexander-cit committed:
[CL-2535] Fix CR comments

alexander-cit committed:
[CL-2535] Rename local variable

alexander-cit commented:
Fixed the comment about variable `projectCardImage` names and new `cardImageWasChanged` variable. But I couldn't add a similar `cardImageWasRemoved` variable because TS was complaining
```ts
        const cardImageWasRemoved =
          projectCardImageToRemove &&
          projectCardImageToRemove.remote &&
          projectCardImageToRemove.id &&
          isString(projectCardImageToRemove.id);
        const cardImageToRemovePromise =
          cardImageWasRemoved && latestProjectId
            ? deleteProjectImage(latestProjectId, projectCardImageToRemove.id)
```
```
const projectCardImageToRemove: UploadFile
Argument of type 'string | undefined' is not assignable to parameter of type 'string'.
  Type 'undefined' is not assignable to type 'string'.ts(2345)
```

alexander-cit commented:
@brentguf Is it fine to merge?

brentguf commented:
I'll review tomorrow morning. You can re-request a review when it's ready to be checked again.

brentguf commented on front/app/containers/Admin/projects/project/general/index.tsx:
```diff
@@ -290,24 +310,24 @@ const AdminProjectsProjectGeneral = () => {
+          isString(projectCardImageToRemove.id) &&
+          latestProjectId
+            ? deleteProjectImage(latestProjectId, projectCardImageToRemove.id)
+            : null;
```
How about this?

```
const cardImageToRemovePromise =
          projectCardImageToRemove?.id && latestProjectId
            ? deleteProjectImage(latestProjectId, projectCardImageToRemove.id)
            : null;
```

You don't need the remote check normally.

alexander-cit replied:
I didn't want to remove the previously existing checks, but maybe it can be done.

brentguf replied:
I can't think of why we would need the remote check. We btw already check for it when we set the `projectCardImageToRemove`.

```
    projectCardImageToRemove.remote &&
      setProjectCardImageToRemove(projectCardImageToRemove);
```

brentguf reviewed (approved):
Nothing blocking. I think you could simplify some variables though. It'll reduce time spent understanding in the future ("why is this being checked here?").

alexander-cit committed:
[CL-2535] [CL-2644] Get rid of (probably) unneeded conditions

brentguf commented:
Good to merge 👍
